cpp/harvest/nacjd_data_publication.cc

```diff
@@ -351,28 +365,79 @@ void ExtractInfoFromNACJD(const std::string &json_path, std::vector<NacjdDoc> *
             nacjd_doc.publisher_ = doc.at("PUBLISHER").get<std::string>();
         }
         if (doc.contains("authors_splitSPLIT")) {
-            nacjd_doc.authors_split = doc.at("authors_splitSPLIT").get<std::vector<std::string>>();
+            nacjd_doc.authors_split = doc.at("authors_splitSPLIT").get<std::set<std::string>>();
         }
         if (doc.contains("STUDYTITLE")) {
-            nacjd_doc.study_titles_ = doc.at("STUDYTITLE").get<std::vector<std::string>>();
+            nacjd_doc.study_titles_ = doc.at("STUDYTITLE").get<std::set<std::string>>();
         }
 
 
         nacjd_docs->emplace_back(nacjd_doc);
     }
 }
 
+void BuildISSNCache(std::map<std::string, std::string> * const issn_cache, const std::string &source_file_name) {
+    auto input_file(MARC::Reader::Factory(source_file_name));
+
+    while (MARC::Record record = input_file->read()) {
+        std::string tag_001;
+        for (auto &field : record) {
+            if (field.getTag() == "001")
+                tag_001 = field.getContents();
+
+            if (field.getTag() == "022") {
+                std::string tag_022 = StringUtil::ASCIIToUpper(field.getFirstSubfieldWithCode('a'));
+                if (not tag_022.empty()) {
+                    issn_cache->insert(std::pair(tag_022, "(DE-627)" + tag_001));
+                }
+            }
+        }
+    }
+}
+
+void AugmentingPPN(std::map<std::string, std::string> const &issn_cache, std::string const &input_file_name,
+                   std::string const &output_file_name) {
+    auto input_file(MARC::Reader::Factory(input_file_name));
+    auto output_file(MARC::Writer::Factory(output_file_name));
+
+    while (MARC::Record record = input_file->read()) {
+        for (auto &field : record) {
+            MARC::Subfields subfields(field.getSubfields());
+            if (field.getTag() == "773") {
+                const std::string issn(StringUtil::ASCIIToUpper(field.getFirstSubfieldWithCode('x')));
+                // std::map<std::string, std::string>::iterator augment_info = issn_cache.find(issn);
+
+                if (issn_cache.find(issn) != issn_cache.end()) {
+                    if (!subfields.replaceFirstSubfield('w', issn_cache.find(issn)->second)) {
+                        subfields.addSubfield('w', issn_cache.find(issn)->second);
+                        std::cout << "Writing new ppn" << issn_cache.find(issn)->second;
+                    }
+                    field.setSubfields(subfields);
+                }
+            }
+        }
+        output_file->write(record);
+    }
+}
 } // unnamed namespace
 
 int Main(int argc, char **argv) {
-    if (argc < 2)
+    if (argc < 3)
         Usage();
 
-    // const std::string mode(argv[1]);
+    const std::string mode(argv[1]);
+    if (mode == "construct") {
+        std::vector<NacjdDoc> nacjd_docs;
+        ExtractInfoFromNACJD(argv[2], &nacjd_docs);
+        WriteMarc(argv[3], nacjd_docs);
+    } else if (mode == "augment") {
```

Comment:
As adding the PPNs is integral to the entire process, it can be done as part of the generation process.